File: src/main/java/plugin/interaction/combat/special/range/MagicBowSpecialAttackPlugin.java
package plugin.interaction.combat.special.range;

import org.redrune.game.content.entity.actor.combat.player.AbstractCombatStyle;
import org.redrune.game.entity.actor.Actor;
import org.redrune.game.entity.actor.mask.Animation;
import org.redrune.game.entity.actor.player.Player;
import org.redrune.game.content.plugin.combat.RangeWeaponPlugin;
import org.redrune.game.content.plugin.combat.SpecialAttackPlugin;
import org.redrune.game.entity.projectile.Projectile;
import org.redrune.game.entity.projectile.ProjectileManager;

/**
 * @author Tyluur <itstyluur@gmail.com>
 * @since 9/6/2017
 */
public class MagicBowSpecialAttackPlugin extends SpecialAttackPlugin {
	
	private static final Animation ANIMATION = new Animation(1074);
	
	@Override
	public int[] getWeaponIds() {
		return arguments(859, 861, 10284, 18332);
	}
	
	@Override
	public void fire(Player source, Actor target, AbstractCombatStyle style) {
		source.setNextAnimation(ANIMATION);
		
		style.sendHit(source, target, style.getCalculator().getMaximumHit(source, 1), style.getRandomDamage(source, target, 1), ProjectileManager.getProjectileDelay(source, target));
		style.sendHit(source, target, style.getCalculator().getMaximumHit(source, 1), style.getRandomDamage(source, target, 1), ProjectileManager.getProjectileDelay(source, target));
		
		visualize(source, target);
		
		RangeWeaponPlugin.dropAmmo(source, target, 1);
		RangeWeaponPlugin.dropAmmo(source, target, 1);
	}
	
	/**
	 * Visualizes the projectiles
	 *
	 * @param source
	 * 		The projectile from
	 * @param target
	 * 		The projectile to
	 */
	private void visualize(Player source, Actor target) {
		int speed = (int) (27.0D + source.getDistance(target) * 5.0D);
		ProjectileManager.sendProjectile(new Projectile(source, target, 249, 41, 36, 20, speed, 15, 0));
		speed = (int) (20.0D + source.getDistance(target) * 10.0D);
		ProjectileManager.sendProjectile(new Projectile(source, target, 249, 41, 36, 40, speed, 10, 0));
	}
}

File: src/main/java/plugin/interaction/combat/special/range/MagicBowSpecialAttackPlugin.kt
package plugin.interaction.combat.special.range

import game.content.plugin.combat.SpecialAttackPlugin
import game.entity.actor.Actor
import game.content.entity.actor.combat.player.AbstractCombatStyle
import game.entity.actor.mask.Animation
import game.entity.projectile.ProjectileManager
import game.content.plugin.combat.RangeWeaponPlugin
import game.entity.projectile.Projectile
import game.entity.actor.player.Player

/**
 * @author Tyluur <itstyluur></itstyluur>@icloud.com>
 * @since 9/6/2017
 */
class MagicBowSpecialAttackPlugin : SpecialAttackPlugin() {
    override fun getWeaponIds(): IntArray {
        return arguments(859, 861, 10284, 18332)
    }

    override fun fire(source: Player, target: Actor, style: AbstractCombatStyle) {
        source.nextAnimation = ANIMATION
        style.sendHit(
            source,
            target,
            style.calculator.getMaximumHit(source, 1.0),
            style.getRandomDamage(source, target, 1.0),
            ProjectileManager.getProjectileDelay(source, target)
        )
        style.sendHit(
            source,
            target,
            style.calculator.getMaximumHit(source, 1.0),
            style.getRandomDamage(source, target, 1.0),
            ProjectileManager.getProjectileDelay(source, target)
        )
        visualize(source, target)
        RangeWeaponPlugin.dropAmmo(source, target, 1)
        RangeWeaponPlugin.dropAmmo(source, target, 1)
    }

    /**
     * Visualizes the projectiles
     *
     * @param source
     * The projectile from
     * @param target
     * The projectile to
     */
    private fun visualize(source: Player, target: Actor) {
        var speed = (27.0 + source.getDistance(target) * 5.0).toInt()
        ProjectileManager.sendProjectile(Projectile(source, target, 249, 41, 36, 20, speed, 15, 0))
        speed = (20.0 + source.getDistance(target) * 10.0).toInt()
        ProjectileManager.sendProjectile(Projectile(source, target, 249, 41, 36, 40, speed, 10, 0))
    }

    companion object {
        private val ANIMATION = Animation(1074)
    }
}